File: recursion/binary_search_recursive.rb
def binary_search(array, left, right, target)
  if left + 1 >= right
    return left if array[left] == target
    return right if array[right] == target
    return -1
  end

  mid = left + (right - left) / 2

  if array[mid] < target
    binary_search(array, mid, right, target)
  else
    binary_search(array, left, mid, target)
  end
end

puts binary_search([0, 1, 2, 5, 7], 0, 4, 5) == 3
puts binary_search([0, 1, 2, 5, 7], 0, 4, 6) == -1
puts binary_search([0, 1, 2, 5, 7], 0, 4, 0) == 0
puts binary_search([0, 1], 0, 1, 1) == 1
puts binary_search([0], 0, 0, 0) == 0

File: recursion/binary_search_recursive.js
var binarySearch = function(array, left, right, target) {
  if (left + 1 >= right) {
    if (array[left] === target) { return left; }
    if (array[right] === target) { return right; }
    return -1;
  }

  var mid = Math.floor(left + (right - left) / 2);

  if (array[mid] === target) {
    return mid;
  } else if (array[mid] < target) {
    return binarySearch(array, mid, right, target);
  } else {
    return binarySearch(array, left, mid, target);
  }
};

console.log(binarySearch([0, 2, 4, 5, 7, 8, 10], 0, 6, 7) === 4);
console.log(binarySearch([0, 2, 4, 5, 7, 8, 10], 0, 6, 0) === 0);
